pkg/service/jetstream.go

```diff
@@ -0,0 +1,212 @@
+	return base64.RawStdEncoding.EncodeToString(sum[:])
+}
+
+func (b *Service) jsStreamName(hash string) string {
+	return fmt.Sprintf("%s-stream-%s", b.Name, hash)
+}
+
+func (b *Service) jsConsumerName(hash string) string {
+	return fmt.Sprintf("%s-%s-%s", b.Identity, b.Name, hash)
+}
+
+// AddStream is an experimental feature that creates a durable stream. It is possible to
+// subscribe to this durable stream using regular Subscribe or SubscribeTo methods given that
+// the subject is included in the created stream.
+// Stream names must be explicit(no pattern matching) and must belong to only one stream.
+//
+// The interface for this feature is experimental and it should be expected to change.
+//
+// NOTE: Messages are automatically acknowledged after handler returns.
+func (b *Service) AddStream(maxMsgs, maxBytes uint64, age time.Duration, subjects ...string) error {
+	if b.js == nil {
+		return ErrNotAvailable
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	for _, s := range subjects {
+		if strings.ContainsAny(s, ">*") {
+			return fmt.Errorf("%s unsupported", s)
+		}
+
+		if _, ok := b.streamSubjects[s]; ok {
+			return fmt.Errorf("%s already configured", s)
+		}
+	}
+
+	hash := b.jsMakeHash(subjects...)
+	streamName := b.jsStreamName(hash)
+
+	connCtx, connCancelFn := context.WithTimeout(b.Context, 10*time.Second)
+	defer connCancelFn()
+
+	cfg := &nats.StreamConfig{
+		Name:     streamName,
+		Subjects: subjects,
+		MaxMsgs:  int64(maxMsgs),
+		MaxBytes: int64(maxBytes),
+		MaxAge:   age,
+	}
+	si, err := b.js.AddStream(cfg, nats.Context(connCtx))
+	if err != nil {
+		return fmt.Errorf("AddStream failed: %w", err)
+	}
+
+	ccfg := &nats.ConsumerConfig{
+		Durable:       b.jsConsumerName(hash),
+		AckPolicy:     nats.AckExplicitPolicy,
+		DeliverPolicy: nats.DeliverAllPolicy,
+	}
+	ci, err := b.js.AddConsumer(streamName, ccfg)
+	if errors.Is(err, nats.ErrConsumerNameAlreadyInUse) {
+		ci, err = b.js.UpdateConsumer(streamName, ccfg)
+	}
+	if err != nil {
+		return fmt.Errorf("AddConsumer failed: %w", err)
+	}
+
+	b.streams = append(b.streams, jsStream{
+		cfgStream:    cfg,
+		cfgConsumer:  ccfg,
+		streamInfo:   si,
+		consumerInfo: ci,
+	})
+
+	for _, s := range subjects {
+		b.streamSubjects[s] = len(b.streams) - 1
+	}
+
+	return nil
+}
+
+// RemoveStream will attempt to remove consumers and streams based on a list of subjects.
+// List of subjects must be exactly the same as was used in AddStream since js Stream and js Consumer names
+// are based on the subjects.
+func (b *Service) RemoveStream(subjects ...string) error {
+	if b.js == nil {
+		return ErrNotAvailable
+	}
+
+	hash := b.jsMakeHash(subjects...)
+	streamName := b.jsStreamName(hash)
+	consumerName := b.jsConsumerName(hash)
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	ctx, cancel := context.WithTimeout(b.Context, 30*time.Second)
+	defer cancel()
+
+	err := b.js.DeleteConsumer(streamName, consumerName, nats.Context(ctx))
```

Comment:
This step is not necessary, as all consumer assigned to the stream will be removed automatically once stream is removed.

Reply:
Got it. I noticed it when I used nats cli and still thought it was a good idea to clean it the reverse way.